dataplane/handlers/interface.go

```diff
@@ -207,67 +207,105 @@ func (ni *Interface) reconcile(config *oc.Interface) {
+			if err := kernel.SetInterfaceHWAddr(config.GetName(), state.GetEthernet().GetHwMacAddress()); err != nil {
+				log.Warningf("Failed to set mac address of port: %v", err)
 			}
 		}
 	}
+
 	if config.GetOrCreateSubinterface(0).Enabled != nil {
 		if state.GetOrCreateSubinterface(0).Enabled == nil || config.GetSubinterface(0).GetEnabled() != state.GetSubinterface(0).GetEnabled() {
 			log.V(1).Infof("setting interface %s enabled %t", tapName, config.GetSubinterface(0).GetEnabled())
 			if err := kernel.SetInterfaceState(tapName, config.GetSubinterface(0).GetEnabled()); err != nil {
 				log.Warningf("Failed to set state address of port: %v", err)
-				return
 			}
 		}
 	}
-	// TODO: refactor this.
+
+	type prefixPair struct {
+		cfgIP, stateIP *string
+		cfgPL, statePL *uint8
+	}
+
+	// Get all state IPs and their corresponding config IPs (if they exist).
+	var interfacePairs []*prefixPair
+	for _, addr := range state.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
+		pair := &prefixPair{
+			stateIP: addr.Ip,
+			statePL: addr.PrefixLength,
+		}
+		if pairAddr := config.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp()); pairAddr != nil {
+			pair.cfgIP = pairAddr.Ip
+			pair.cfgPL = pairAddr.PrefixLength
+		}
+		interfacePairs = append(interfacePairs, pair)
+	}
+	for _, addr := range state.GetOrCreateSubinterface(0).GetOrCreateIpv6().Address {
+		pair := &prefixPair{
+			stateIP: addr.Ip,
+			statePL: addr.PrefixLength,
+		}
+		if pairAddr := config.GetSubinterface(0).GetIpv6().GetAddress(addr.GetIp()); pairAddr != nil {
+			pair.cfgIP = pairAddr.Ip
+			pair.cfgPL = pairAddr.PrefixLength
+		}
+		interfacePairs = append(interfacePairs, pair)
+	}
+
+	// Get all config IPs and their corresponding state IPs (if they exist).
 	for _, addr := range config.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
-		configIP := addr.Ip
-		configPL := addr.PrefixLength
-		var stateIP *string
-		var statePL *uint8
-		stateAddr := state.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp())
-		if stateAddr != nil {
-			stateIP = stateAddr.Ip
-			statePL = stateAddr.PrefixLength
-		}
-
-		if configIP != nil && configPL != nil && (stateIP == nil || *statePL != *configPL) {
-			log.V(1).Infof("Config IP: %v, Config PL: %v. State IP: %v, State PL: %v", addr.GetIp(), addr.GetPrefixLength(), stateAddr.GetIp(), stateAddr.GetPrefixLength())
-			log.V(2).Infof("setting interface %s ip %s/%d", tapName, *configIP, *configPL)
-			if err := kernel.SetInterfaceIP(tapName, *configIP, int(*configPL)); err != nil {
-				log.Warningf("Failed to set ip address of port: %v", err)
-				return
-			}
+		pair := &prefixPair{
+			cfgIP: addr.Ip,
+			cfgPL: addr.PrefixLength,
+		}
+		if pairAddr := state.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp()); pairAddr != nil {
+			pair.stateIP = pairAddr.Ip
+			pair.statePL = pairAddr.PrefixLength
 		}
+		interfacePairs = append(interfacePairs, pair)
```

Comment:
I guess for the second pair of for loops you should only add if the state doesn't exist, as otherwise you have a duplicate entry.

Reply:
we could have the same state IP but a different prefix length, so we need to add it. the duplicate entry shouldn't cause an issue